### PR TITLE
fix: [DHIS2-2652] Block deletion of org unit group and set linked to …

### PR DESCRIPTION
**Pull request by enricocolasante**
…analytics objects

I am just not sure about Chart.

Should I create a deletion handler for it?

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventchart/EventChartDeletionHandler.java [hunk 1]
```diff
@@ -129,13 +137,36 @@ public void deleteProgramStage( ProgramStage programStage )
...
             {
                 eventChartService.deleteEventChart( chart );
             }
         }
     }
+
+    @Override
+    public String allowDeleteOrganisationUnitGroupSet( OrganisationUnitGroupSet groupSet )
+    {
+        String sql = "select count(*) "
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventchart/EventChartDeletionHandler.java:153]: For all your SQL, stick with either uppercase or lowercase for SQL keywords :)

Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventchart/EventChartDeletionHandler.java:153]: Also, the WHERE on a new line maybe :)?

**Comment by stian-sandvold** [comment]
We should block also if it is chart. You can create a new deletionhandler for that :)

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventchart/EventChartDeletionHandler.java [hunk 2]
```diff
@@ -129,13 +137,37 @@ public void deleteProgramStage( ProgramStage programStage )
...
+            + groupSet.getId();
+
+        return jdbcTemplate.queryForObject( sql, Integer.class ) == 0 ? null : "orgunitgroupsetdimension";
+    }
+
+    @Override
+    public String allowDeleteOrganisationUnitGroup( OrganisationUnitGroup group )
+    {
+        String sql = "SELECT count(*) "
+            + " FROM orgunitgroupsetdimension d JOIN eventchart_orgunitgroupsetdimensions e "
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventchart/EventChartDeletionHandler.java:166]: Use explicit join ("inner join") to remove ambiguity.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventreport/EventReportDeletionHandler.java [hunk 3]
```diff
@@ -129,13 +138,36 @@ public void deleteProgramStage( ProgramStage programStage )
...
+                eventReportService.deleteEventReport( chart );
             }
         }
     }
+
+    @Override
+    public String allowDeleteOrganisationUnitGroupSet( OrganisationUnitGroupSet groupSet )
+    {
+        String sql = "SELECT count(*) "
+            + " FROM orgunitgroupsetdimension d JOIN eventreport_orgunitgroupsetdimensions e "
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventreport/EventReportDeletionHandler.java:155]: Use lower-case only in SQL statements for consistency.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/GenericAnalyticalObjectDeletionHandler.java [hunk 4]
```diff
@@ -96,10 +106,52 @@ public String allowDeletePeriod( Period period )
...
+        removeItem( getAnalyticalObjectService().getAnalyticalObjects( organisationUnit ), organisationUnit,
+            ( ao, di ) -> ao.getOrganisationUnits().remove( di ) );
+    }
+
+    @Override
+    public void deleteOrganisationUnitGroup( OrganisationUnitGroup organisationUnitGroup )
+    {
+        removeItem( getAnalyticalObjectService().getAnalyticalObjects( organisationUnitGroup ), organisationUnitGroup,
+            ( ao, di ) -> {
+                List<OrganisationUnitGroupSetDimension> dimensionsToDelete = ao.getOrganisationUnitGroupSetDimensions()
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/GenericAnalyticalObjectDeletionHandler.java:118]: From the store we are retrieving AnalyticalObjects with the provided dimension.
In the consumer we are deleting the dimension linked to the group that we are deleting.
So we know that in the AnalyticalObject there is something to delete, but we do not want to delete all the dimensions in it.

[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/GenericAnalyticalObjectDeletionHandler.java:118]: Right, makes sense, got it.

**Pull request merged**